src/lib/autoDetect.ts

```diff
@@ -0,0 +1,106 @@
+import {
+  ExtensionContext,
+  workspace,
+  window,
+  ConfigurationTarget,
+  WorkspaceConfiguration,
+} from 'vscode';
+import { FileAssociations } from './types';
+const { extname } = require('path');
+import {
+  EXTENSION_CONFIG_NAME,
+  EXTENSION_CONFIG_KEYS,
+  HUBL_HTML_ID,
+  HUBL_CSS_ID,
+} from './constants';
+
+export const initializeHubLAutoDetect = (context: ExtensionContext) => {
+  if (
+    hasHubLAssociations() ||
+    checkGlobalState() ||
+    !!context.workspaceState.get('DO_NOT_USE_HUBL')
+  ) {
+    return;
+  }
+  const showHublDetectedMessage = () => {
+    const hublDetectedMessage = `It looks like this file contains HubL code. Would you like to use HubL for this project?`;
+    const hublDetectedYesButton = `Use HubL`;
+    const hublDetectedNoButton = 'No';
+    const hublDetectedNeverAgainButton = 'Never ask again';
+
+    window
+      .showInformationMessage(
+        hublDetectedMessage,
+        hublDetectedYesButton,
+        hublDetectedNoButton,
+        hublDetectedNeverAgainButton
+      )
+      .then((selection) => {
+        switch (selection) {
+          case hublDetectedYesButton: {
+            updateWorkspaceFileAssociation();
+          }
+          case hublDetectedNoButton: {
+            context.workspaceState.update('DO_NOT_USE_HUBL', true);
+          }
+          case hublDetectedNeverAgainButton: {
+            workspace
+              .getConfiguration(EXTENSION_CONFIG_NAME)
+              .update(
+                EXTENSION_CONFIG_KEYS.NEVER_USE_HUBL,
+                true,
+                ConfigurationTarget.Global
+              );
+          }
+          default: // User closed the dialogue
+        }
+        handleTextDocumentOpen.dispose();
+      });
+  };
+
+  const handleTextDocumentOpen = workspace.onDidOpenTextDocument(() => {
+    const textDocument = window.activeTextEditor?.document;
+    if (!textDocument) {
+      return;
+    }
+    const fileExt = extname(textDocument.fileName);
+    if (fileExt !== '.html' && fileExt !== '.css') {
+      return;
+    }
+
+    let n = 1;
+    const limit = textDocument.lineCount;
+    while (n < 50 && n <= limit) {
+      const lineWithHubl = textDocument.lineAt(n).text.match(/{{.*}}|{%.*%}/g);
+      if (lineWithHubl) {
+        showHublDetectedMessage();
+        break;
+      }
+      n++;
+    }
+  });
+};
```

Comment:
Not technically necessary but I'm a [SRP ](https://en.wikipedia.org/wiki/Single-responsibility_principle) evangelist - you could still break these functions up a bit more. If you move the listener disposal to the right after the `showHublDetectedMessage()` (same place effectively, since it's at the end of all the prompt responses), you can separate out the prompting function and the listener, only passing them the context, and leaving the "main" `initializeHubLAutoDetect` to just checking those conditions and then starting the detection if none apply. Very pretty! 

<img width="649" alt="Screenshot 2023-07-28 at 3 26 48 PM" src="https://github.com/HubSpot/hubspot-cms-vscode/assets/25852903/251bf92d-62a1-44de-9917-f2c5b71e10a4">